### PR TITLE
Serve http2 when listener wrapper doesn't return *tls.Conn

### PR DESCRIPTION
**Pull request by WeidiDeng**
fixes [4921](https://github.com/caddyserver/caddy/issues/4921)

### CODE REVIEW

**Comment by mholt** [comment]
Thanks for working on this!

So, my understanding of this is, that it allows us to serve HTTP/2 even if the underlying Conn is wrapped by a ListenerWrapper plugin so that if it's not a tls.Conn, it can still be recognized as one?

**Comment by WeidiDeng** [comment]
> Thanks for working on this!
> 
> So, my understanding of this is, that it allows us to serve HTTP/2 even if the underlying Conn is wrapped by a ListenerWrapper plugin so that if it's not a tls.Conn, it can still be recognized as one?

Yes, because golang http.server won't directly serve http2 when return non tls.Conn, I have to manually configure [h2server](https://github.com/golang/go/issues/12737). tls.Conn still uses http.Server code path, because when handshake is not complete, it won't have "h2" as negotiated protocol.

**Comment by WeidiDeng** [comment]
@mholt I noticed you don't allow serve [only http2](https://github.com/caddyserver/caddy/blob/c7772588bd44ceffcc0ba4817e4d43c826675379/modules/caddyhttp/app.go#L189), if you want, I can use http2.Server.ServeConn in this case.

**Comment by mholt** [comment]
@WeidiDeng Yeah, we'd have to write our own HTTP server. I'm not sure I'm quite ready for that since it seems like a maintenance burden. You can see my attempt here: https://github.com/caddyserver/caddy/pull/4707#issuecomment-1204438638

I'll review this soon!

**Comment by WeidiDeng** [comment]
@mholt I looked at you code, that's because http2.Server requires tls handshake to be complete, if there is any. To be fair, http2.Server can serve h2c with prior knowledge, that's like how currently reverse_proxy h2c transport is implemented.

But yes I agree it's a burden, especially if we are allowing serve http2 with non tls.Conn. Perhaps it's easier to configure tls nextproto directly, normal clients won't send h1 requests on a h2 tls.Conn.

**Comment by francislavoie** [comment]
@WeidiDeng could you rebase this?

I'm not sure there's anything really blocking merging this. Seems like a reasonable fix for a usecase only used by caddy-l4 currently.

**Comment by WeidiDeng** [comment]
@francislavoie I'll also need to update l4 listener module, that part is left as a todo.

There are some parts that I'm not sure, like using x/net/http2 instead of builtin http2, should there be an option?

File: modules/caddyhttp/server.go [hunk 1]
```diff
@@ -213,6 +214,14 @@ type Server struct {
 
 // ServeHTTP is the entry point for all HTTP requests.
 func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.TLS == nil {
```
Comment (mholt) [inline modules/caddyhttp/server.go:217]: Maybe a comment explaining what this block is used for

File: modules/caddyhttp/server.go [hunk 2]
```diff
@@ -195,10 +195,11 @@ type Server struct {
...
 	ctx          caddy.Context
 
-	server      *http.Server
-	h3server    *http3.Server
-	h3listeners []net.PacketConn // TODO: we have to hold these because quic-go won't close listeners it didn't create
-	addresses   []caddy.NetworkAddress
+	server         *http.Server
+	h3server       *http3.Server
+	h3listeners    []net.PacketConn // TODO: we have to hold these because quic-go won't close listeners it didn't create
+	http2listeners []*http2Listener
```
Comment (mholt) [inline modules/caddyhttp/server.go:201]:
```suggestion
	h2listeners []*http2Listener
```

Just for symmetry with `h3` vars I guess :)

File: modules/caddyhttp/http2listener.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	weakrand "math/rand"
+	"net"
+	"net/http"
+	"sync/atomic"
+	"time"
+
+	"golang.org/x/net/http2"
+)
+
+type http2Listener struct {
```
Comment (mholt) [inline modules/caddyhttp/http2listener.go:15]: A comment explaining the usefulness of this type instead of just the http2 package types would be useful!

**Review by mholt — CHANGES_REQUESTED** [review]
Thanks, I made a first pass and have some minor suggestions -- This looks like we're doing a lot more in-house but it opens up some new possibilities, yeah?

**Review by mholt — APPROVED** [review]
Thanks @WeidiDeng  -- let's give this a try. Sorry for the long time on this. Thank you for your work!

**Pull request merged**